### PR TITLE
Update the Android project template settings

### PR DESCRIPTION
**Pull request by felipecsl**
The Android template project was using reeeeally outdated dependencies.
This PR bumps them to saner and more recent versions.

### CODE REVIEW

File: local-cli/generator-android/templates/src/app/build.gradle [hunk 1]
```diff
@@ -83,13 +83,13 @@ def enableSeparateBuildPerCPUArchitecture = false
...
-    compileSdkVersion 23
-    buildToolsVersion "23.0.1"
+    compileSdkVersion 24
+    buildToolsVersion "24.0.2"
 
     defaultConfig {
         applicationId "<%= package %>"
         minSdkVersion 16
-        targetSdkVersion 22
+        targetSdkVersion 24
```
Comment (satya164) [inline local-cli/generator-android/templates/src/app/build.gradle]: Changing this will crash apps when a permission is denied by the user. Actually none of the inbuilt modules like geolocation are asking the permission for location. So this will likely break a lot of things.

**Comment by satya164** [comment]
Have you tested this? It might not be as simple as bumping the versions.

cc @bestander

File: local-cli/generator-android/templates/src/app/build.gradle [hunk 2]
```diff
@@ -127,7 +127,7 @@ android {
 
 dependencies {
     compile fileTree(dir: "libs", include: ["*.jar"])
-    compile "com.android.support:appcompat-v7:23.0.1"
+    compile "com.android.support:appcompat-v7:24.2.0"
```
Comment (bestander) [inline local-cli/generator-android/templates/src/app/build.gradle]: Updating appcompat version is not that easy.
RN is coupled with 23.0.1 right now because we have Java code in RN that is coupled to specific versions of its dependencies, see my [explanation](https://github.com/facebook/react-native/pull/9340).

The best way to move forward is if someone abstracts out RN dependencies so that users could provide/switch implementations during build times.
Help is wanted and needed.

**Comment by bestander** [comment]
Bumping Gradle only seems fairly safe, bumping libs versions need a justification and extra work.

Please mind that for RN source code build we are going to replace Gradle with Buck soon https://www.facebook.com/groups/reactnativeoss/permalink/1580665485563413/

**Comment by felipecsl** [comment]
Yep I tested this and the template project works with these settings. 
As for the dependency versions, I understand the hassle. Thanks for the explanation. 
@bestander I can probably take that. My plan is to create a shim interface for the places where RN depends on the support library code and have that be pluggable by the user if needed. Makes sense? I can draft something today

**Comment by satya164** [comment]
@felipecsl Last time I tried, it was crashing when trying to access geolocation, since it doesn't handle this case. Did it work for you?

**Comment by felipecsl** [comment]
Yeah the sample project worked fine, but using any geolocation features would probably throw SecurityExceptions

**Comment by MajeedThaikas** [comment]
bump up

**Comment by MajeedThaikas** [comment]
bump up

**Comment by felipecsl** [comment]
We won't be able to merge this as-is. I started working on the shims for the appcompat dependencies, but that's not straightforward and will take a few more days.

**Comment by bestander** [comment]
Yes, @felipecsl, that sounds great, please proceed :)

**Comment by felipecsl** [comment]
@bestander I spent a some time drafting the appcompat dependencies shim work but that's actually still not clear to me exactly how that would work. The way I see it right now, it will be a really big amount of work and not sure it's worth it at all, but maybe I'm just not picturing it correctly.
Would you mind outlining the high level steps you're seeing for how this would work to make sure we're on the same page?

**Comment by bestander** [comment]
I feel your concerns, @felipecsl. 

I was thinking that decoupling appcompat should be feasible with a few gotchas.

Here is what I would be doing:

1. in Java we have 4 places where appcompat library is imported: https://github.com/facebook/react-native/search?utf8=%E2%9C%93&q=%22import+android.support.v7%22&type=Code

Could this be replaced with a dummy interface, e.g. create com.facebook.react.app-compat that mirrors the classes that RN uses from Android AppCompat, or does it pull through a whole tree of interfaces and it involves too much code that needs to be duplicated?

2. The implementations of the dummy interface could be in folders com.facebook.react.app-compat.v23_1_0 and com.facebook.react.app-compat.v23_0_1 that would work as bridge between React Native and specific versions of appcompat.

3. The tricky part is to make the build system switch between v23_1_0 and v23_0_1 implementations with a configuration parameter.
The had part is to make it work with Buck, Gradle and both for RN main repo and RN apps that depend on RN.
This would require a discussion and maybe a compromise (e.g. only Gradle and only RN apps can switch appcompat version while all others will have a hardcoded v23_0_1)

Is it the path you were trying out?

**Comment by felipecsl** [comment]
@bestander yep, that's exactly what I was trying.

However, using a dummy interface for `RecyclerView` means that you also have to replace all its inner classes, like `RecyclerViewDataObserver`, `Adapter`, `LayoutManager`, `ItemDecoration` and many others. That involves also proxying all the methods that interact with these methods. After 30min-1hour of fiddling it was quite clear that it was growing out of control and becoming a monster 😄 

tl;dr: I don't think it's worth it unless we can significantly reduce the touch points.

**Comment by bestander** [comment]
Yeah, makes sense, I thought the same.
Thanks for trying anyway

**Comment by bestander** [comment]
This is quite a power feature, probably forking the framework for people would be easier than supporting multiple android sdk libraries.

**Comment by jpshelley** [comment]
For those interested, we have a fork that could be of use once Jitpack starts building it, or you can clone and create the .aar yourself https://github.com/facebook/react-native/pull/9340#issuecomment-266092481
